staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/escaping_test.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"fmt"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	"k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"testing"
+)
+
+func TestPropNameEscaping(t *testing.T) {
+	cases := sets.NewString("self", "_if", "__if", "___if").Union(model.AlwaysReservedIdentifiers).Union(model.RootReservedIdentifiers)
+	for _, prop := range cases.List() {
+		escapedProp := model.Escape(prop)
+		scopedIfNonRootProp := escapedProp
+		if model.IsRootReserved(prop) || prop == ScopedVarName {
+			scopedIfNonRootProp = "self." + scopedIfNonRootProp
+		}
+		t.Run(prop, func(t *testing.T) {
+			s := &schema.Structural{
+				Generic: schema.Generic{
+					Type: "object",
+				},
+				Properties: map[string]schema.Structural{
+					// prop names are not escaped in the schema
+					prop: {
+						Generic: schema.Generic{
+							Type: "array",
+						},
+						Items: &schema.Structural{
+							Generic: schema.Generic{
+								Type: "string",
+							},
+						},
+					},
+				},
+				Extensions: schema.Extensions{
+					XValidations: apiextensions.ValidationRules{
+						{
+							// prop names are escaped in CEL validation rules
+							Rule: fmt.Sprintf("has(self.%s) && %s == ['a', 'b', 'c'] && size(%s.filter(e, e == 'c')) == 1", escapedProp, scopedIfNonRootProp, scopedIfNonRootProp),
```

Comment:
Is part of the motivation for the escapes to make sure that the `has` macro can be used with map-like key names?

Reply:
Yeah, and more generally support field selection for all valid property names. The most problematic reserved CEL lexer keyword is 'namespace' since that is a commonly used property name in Kubernetes.  The escaping rule means developers can do things like `spec.something._namespace` to refer to the namespace field. (Correct me if I'm wrong, but my reading of the BNF for the lexer is that `spec.something.namespace` wouldn't be legal).

Reply:
(I would love to be able to do backtick keyword escaping like some languages support, i.e. `` spec.something.`namespace` ``, but that would be grammar change).